demo-circuit/src/test_utils.rs

```diff
@@ -0,0 +1,323 @@
+use std::convert::TryInto;
+
+use cctp_primitives::{
+    commitment_tree::CommitmentTree,
+    proving_system::init::{get_g1_committer_key, load_g1_committer_key},
+    type_mapping::{CommitterKeyG1, FieldElement, GingerMHT},
+};
+use primitives::FieldBasedMerkleTree;
+use rand::{rngs::ThreadRng, thread_rng, Rng};
+
+use crate::{
+    sc2sc::{
+        ScCommitmentCertPath, MAX_QUALITY_CERT_HASH_CUSTOM_FIELDS_POS, MSG_MT_HEIGHT,
+        MSG_ROOT_HASH_CUSTOM_FIELDS_POS,
+    },
+    GingerMHTBinaryPath, MAX_SEGMENT_SIZE, SUPPORTED_SEGMENT_SIZE, common::WithdrawalCertificateData,
+};
+
+pub(crate) fn init_g1_committer_key() -> CommitterKeyG1 {
+    let _ = load_g1_committer_key(MAX_SEGMENT_SIZE - 1);
+    let ck_g1 = get_g1_committer_key(Some(SUPPORTED_SEGMENT_SIZE - 1)).unwrap();
+    assert_eq!(ck_g1.comm_key.len(), SUPPORTED_SEGMENT_SIZE);
+    ck_g1
+}
+
+#[derive(Clone)]
+pub(crate) struct RandomWithdrawalCertificateDataBuilder {
+    ledger_id: Option<FieldElement>,
+    epoch_id: Option<u32>,
+    bt_root: Option<FieldElement>,
+    quality: Option<u64>,
+    mcb_sc_txs_com: Option<FieldElement>,
+    ft_min_amount: Option<u64>,
+    btr_min_fee: Option<u64>,
+    custom_fields: Vec<Option<FieldElement>>,
+}
+
+#[allow(unused)]
+impl RandomWithdrawalCertificateDataBuilder {
+    pub(crate) fn new(n_custom_field: usize) -> Self {
+        Self {
+            ledger_id: None,
+            epoch_id: None,
+            bt_root: None,
+            quality: None,
+            mcb_sc_txs_com: None,
+            ft_min_amount: None,
+            btr_min_fee: None,
+            custom_fields: vec![None; n_custom_field],
+        }
+    }
+
+    pub(crate) fn build(&self) -> WithdrawalCertificateData {
+        let mut rng = thread_rng();
+        self.build_with_rng(&mut rng)
+    }
+
+    pub(crate) fn build_with_rng(&self, rng: &mut ThreadRng) -> WithdrawalCertificateData {
```

Comment:
I implemented `Distribution` for the builder `n_custom_field` is not an issue because `sample` is implemented for the builder but return certificates.